File: backend/app/schemas/dish.py
import uuid
from datetime import date, datetime, time
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.models.dish import DishReviewProsConsType, PortionSize, PriceTier


class DishCreate(BaseModel):
    restaurant_id: uuid.UUID
    name: str = Field(max_length=200)
    description: str | None = None
    cover_image_url: str | None = None
    price_tier: PriceTier | None = None


class DishUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    description: str | None = None
    cover_image_url: str | None = None
    price_tier: PriceTier | None = None


class DishResponse(BaseModel):
    id: uuid.UUID
    restaurant_id: uuid.UUID
    name: str
    description: str | None
    cover_image_url: str | None
    price_tier: PriceTier | None
    computed_rating: Decimal
    review_count: int
    created_by: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class DishReviewProsConsCreate(BaseModel):
    type: DishReviewProsConsType
    text: str = Field(max_length=500)


class DishReviewProsConsResponse(BaseModel):
    id: int
    type: DishReviewProsConsType
    text: str

    model_config = {"from_attributes": True}


class DishReviewTagCreate(BaseModel):
    tag: str = Field(max_length=100)


class DishReviewTagResponse(BaseModel):
    id: int
    tag: str

    model_config = {"from_attributes": True}


class DishReviewImageCreate(BaseModel):
    url: str = Field(max_length=500)
    alt_text: str | None = Field(None, max_length=300)
    display_order: int = 0


class DishReviewImageResponse(BaseModel):
    id: uuid.UUID
    url: str
    alt_text: str | None
    display_order: int
    uploaded_at: datetime

    model_config = {"from_attributes": True}


class DishReviewCreate(BaseModel):
    dish_id: uuid.UUID
    date_tasted: date
    time_tasted: time | None = None
    note: str
    rating: int = Field(ge=1, le=5)
    portion_size: PortionSize | None = None
    would_order_again: bool | None = None
    visited_with: str | None = Field(None, max_length=200)
    pros_cons: list[DishReviewProsConsCreate] = []
    tags: list[DishReviewTagCreate] = []
    images: list[DishReviewImageCreate] = []


class DishReviewUpdate(BaseModel):
    date_tasted: date | None = None
    time_tasted: time | None = None
    note: str | None = None
    rating: int | None = Field(None, ge=1, le=5)
    portion_size: PortionSize | None = None
    would_order_again: bool | None = None
    visited_with: str | None = Field(None, max_length=200)


class DishReviewResponse(BaseModel):
    id: uuid.UUID
    dish_id: uuid.UUID
    user_id: uuid.UUID
    user_display_name: str | None = None
    date_tasted: date
    time_tasted: time | None
    note: str
    rating: int
    portion_size: PortionSize | None
    would_order_again: bool | None
    visited_with: str | None
    created_at: datetime
    updated_at: datetime
    pros_cons: list[DishReviewProsConsResponse] = []
    tags: list[DishReviewTagResponse] = []
    images: list[DishReviewImageResponse] = []

    model_config = {"from_attributes": True}

    @model_validator(mode="wrap")
    @classmethod
    def _extract_user_display_name(cls, values, handler):  # type: ignore[no-untyped-def]
        result = handler(values)
        # Extract user display_name from the ORM relationship if available
        if hasattr(values, "user") and values.user is not None:
            result.user_display_name = values.user.display_name
        return result

File: app/lib/types/dish.ts
export type PriceTier = '$' | '$$' | '$$$';
export type PortionSize = 'small' | 'medium' | 'large';
export type DishReviewProsConsType = 'pro' | 'con';
export type PillarScore = 1 | 2 | 3;

export interface Dish {
  id: string;
  restaurant_id: string;
  name: string;
  description: string | null;
  cover_image_url: string | null;
  price_tier: PriceTier | null;
  computed_rating: number;
  review_count: number;
  created_by: string;
  created_at: string;
}

export interface CreateDishRequest {
  name: string;
  description?: string;
  cover_image_url?: string;
  price_tier?: PriceTier;
}

export interface UpdateDishRequest {
  name?: string;
  description?: string | null;
  cover_image_url?: string | null;
  price_tier?: PriceTier | null;
}

export interface DishReviewProsCons {
  id: number;
  type: DishReviewProsConsType;
  text: string;
}

export interface DishReviewTag {
  id: number;
  tag: string;
}

export interface DishReviewImage {
  id: string;
  url: string;
  alt_text: string | null;
  display_order: number;
  uploaded_at: string;
}

export interface DishReview {
  id: string;
  dish_id: string;
  user_id: string;
  user_display_name: string | null;
  date_tasted: string;
  time_tasted: string | null;
  note: string;
  rating: number;
  /** Precio absoluto en la moneda del restaurante. Null cuando el crítico
   * no quiso o no pudo registrarlo (cortesía, menú degustación, reseña antigua). */
  price_paid: number | null;
  portion_size: PortionSize | null;
  would_order_again: boolean | null;
  visited_with: string | null;
  is_anonymous: boolean;
  presentation: PillarScore | null;
  value_prop: PillarScore | null;
  execution: PillarScore | null;
  created_at: string;
  updated_at: string;
  pros_cons: DishReviewProsCons[];
  tags: DishReviewTag[];
  images: DishReviewImage[];
}

export interface CreateReviewRequest {
  date_tasted: string;
  time_tasted?: string;
  note: string;
  rating: number;
  /** Numérico positivo. Omitir cuando no aplica. */
  price_paid?: number;
  portion_size?: PortionSize;
  would_order_again?: boolean;
  visited_with?: string;
  is_anonymous?: boolean;
  presentation?: PillarScore;
  value_prop?: PillarScore;
  execution?: PillarScore;
  pros_cons?: { type: DishReviewProsConsType; text: string }[];
  tags?: { tag: string }[];
  images?: { url: string; alt_text?: string; display_order?: number }[];
}

export interface UpdateReviewRequest {
  date_tasted?: string;
  time_tasted?: string;
  note?: string;
  rating?: number;
  /** Numérico positivo o null para limpiar. */
  price_paid?: number | null;
  portion_size?: PortionSize;
  would_order_again?: boolean;
  visited_with?: string;
  is_anonymous?: boolean;
  presentation?: PillarScore;
  value_prop?: PillarScore;
  execution?: PillarScore;
  /** Renombre del plato. Si el nombre normalizado cambia, el backend re-linkea
   * la review a otro Dish (find-or-create) en el mismo restaurante. */
  dish_name?: string;
  /** When omitted, leave existing rows alone. When [], clear them. */
  pros_cons?: { type: DishReviewProsConsType; text: string }[];
  tags?: { tag: string }[];
  images?: { url: string; alt_text?: string; display_order?: number }[];
}

/** Internal view-model used by restaurant detail sub-components. */
export type Plate = {
  name: string;
  date: string;
  time?: string;
  note: string;
  pros: string[];
  cons: string[];
  image?: string;
  images?: string[];
  rating?: number;
  price?: string;
  portion?: string;
  wouldOrderAgain?: boolean;
  tags?: string[];
  visitedWith?: string;
};
